### DISCUSSION TITLE
Crazyflie, consistent crash

### DISCUSSION THREAD

**Discussion by Kalib-Drift** [bitcraze/discussions#543] [Q&A]
Hey everyone, I am currently having an issue with the crazyflie following what seems like the exact flight path and consistently crashing because of this.

I have my room set up with 4 LPS nodes in TWR mode, tracking the crazyflie decently and accurately. when using the cfclient I have the crazyflie set to hold altitude, when telling it to take off from the ground it will rise then swing to the left then forward into the wall, my first thought was that this is an issue with its motors but I get the same result when testing on different crazyflies. it's funny how accurate it is, I have gotten to a point in my troubleshooting where I tell it to take off and then run across the room to catch it in a specific spot. 

I think it is trying to hold a specific flight path or position outside my area.
how may I tell it to hover at 0,0,0?

**Comment by krichardsson** [comment]
The first question that comes to mind is how your system is set up? How are your 4 anchors placed? 4 anchors could probably work but there is no redundancy and a couple more is generally useful.

The second question is if you place the Crazyflie pointing in the [correct direction, along the x-axis](https://www.bitcraze.io/documentation/tutorials/getting-started-with-flying-using-lps/#restart-the-crazyflie), before taking off

> how may I tell it to hover at 0,0,0?

Check out the `[examples](https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples/autonomy)`

something like this (not complete code)
``` python
    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        with PositionHlCommander(scf) as pc:
            # Hover for 5 seconds
            time.sleep(5)
            # Go to 0, 0, 1
            pc.go_to(0.0, 0.0, 1.0)
```

Note: you probably don't want to go to z=0 as this is the floor

**Reply by Kalib-Drift** [reply]
I have the anchors set up on 4 pillars in my space, 2 opposite corners up top, 2 opposite corners on the bottom

1.(1.68,-1.62,1.75)
2.(1.68, 1.62 , 0)
3.(-1.68, 1.62,1.68)
4. (-1.68, -1.62, 0)
the drone is facing the x positive direction before taking off

I will add some more anchors to this system then check to see if it is already trying to follow a specific flight path in the autonomous sequence script

**Reply by krichardsson** [reply]
The anchor positions looks good to me. I set up a similar system in our lab to see how well it performed. It actually worked pretty well! I had some problems to get a good z-estimate on the floor, but if I moved the CF up a bit it was OK. 

I could fly [the figure 8](https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/autonomy/autonomous_sequence_high_level.py) for instance without problems.

**Reply by Kalib-Drift** [reply]
I added 2 more anchors at (0,-1.62,0.7) and (1.34, 0, 0.77) this seemed to add a bit more stability in the tracking but I'm still running into the same issue.

Running it via cfclient, has the drone take off, swing left, and forward ultimately sending it into a wall.

When running it via python script using motion commander I tell it to take off, move up 1, sleep 6, and stop. It will then rise to an acceptable height but slowly drift forward and to the right. I get a similar result when running a script using a position set telling it to do the same thing with coordinates.

In between flights, I fly using the crazyflie mobile client to check potential mechanical issues or variants. They all fly just fine and seem to fly almost exactly the same way when using both the script and the client.

This makes me think it is still just the tracking system. 

In your big UWB test video, you pointed out you had an issue with the tracking system being too close to another object which was reflecting the signal. My anchors are placed 2cm off 4 metal pillars in my space, could this be the problem?

**Reply by Kalib-Drift** [reply]
I have also tested this with a new crazyflie which had yet to crash and got the same result

**Reply by Kalib-Drift** [reply]
I get the same result in TDOA as well, just with less tracking accuracy

**Reply by krichardsson** [reply]
I'm not sure what the problem might be, with 6 anchors and TWR I would expect positioning to be good. 

> My anchors are placed 2cm off 4 metal pillars in my space, could this be the problem?

It is possible and probably worth investigating. You could try to move them out a bit and see if it improves the situation. 

I think you should make sure the position estimate works well before you try to fly, if the estimate is bad you just risk crashing. 

If you put the CF on a chair or similar in the space, is the position stable? There is always some noise in UWB but it should stay within 5-10 cm. How does it behave in different positions? Does the estimated position match reality?

If you look in the Loco tab in the client, are all the anchors green? I would expect all anchors to be green all the time in such a small space.

You can try to increase the transmission power in the anchors ([connection instructions](https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/development/anchor-low-level-config/)) and the Crazyflie by setting the "Full TX power" build config flag when building the firmware. I don't think is should be necessary as your system is pretty small.

Just to make sure:
* anchor positions are measured in meters (not feet) - there has been some confusion in other cases. I assume this is not the problem here as you system would be tiny if it was measured in feet :-)
* maybe double check that you have configured the correct anchor id at the correct position. There is an "[anchor identification mode](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/loco_tab/#anchor-identification-mode)" you can use to make sure the anchors in the 3D visualization matches reality

**Reply by Kalib-Drift** [reply]
when testing I haven't had any issues with losing anchors while flying, they maintain a green status 100% of the time. 

I am measured in meters, and I have set up and re-set up my room multiple times, it is possible that one of those times I made a mistake but unlikely that I made the same mistake every time, the 3d representation in the loco tab seems accurate to my setup and its easy to tell where the crazyflie is in the room.

The position is quite stable when on the ground (a little off in the z value) or on my workstation, when holding it, there seems to be some sway in the tracking when moving quickly but when moving calmy it seems accurate.

I am currently setting up my space with clamps (taping the anchors to the long arm of the clamps) to suspend the anchors away from the pillars and also make sure they will not move at any point.

I got a relatively good flight test at one point while removing one of the anchors from the system but was unable to replicate it, seemed like that anchor was causing undeeded noise. 

my hope is that by suspending the anchors away from the metal pillars and cement floor i will decrease the sway in the tracking when moving quickly. 

I will look at increasing the anchor's power, alternatively, how might it change its position refresh rate? are they at 100Hz by default?

**Reply by krichardsson** [reply]
`I will look at increasing the anchor's power, alternatively, how might it change its position refresh rate? are they at 100Hz by default?`

In TWR the Crazyflie is ranging with one anchor at a time (round robin), I think it is doing one ranging every 4 ms or there about. The distance data is fed into the kalman estimator that runs at 100 Hz. These rates can not be configured.
The graphics in the client is updated at 1 Hz (I think) but if you want a faster update you can use [the plotter tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/plotter_tab/). Set up a log configuration for logs [`stateEstimate.x`, `stateEstimate.y` and `stateEstimate.z`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#stateestimate)

**Reply by Kalib-Drift** [reply]
after setting up this system again, taking new measurements, and setting up the anchors so they are suspended and secure, it seems to reduce the severity of the issue, but it's still very much present, at this point I'm fairly certain this is simply an issue of the accuracy in the positioning system. 
 
what are some things I can do to increase this accuracy, or what are things that generally cause inaccuracies? What is an acceptable amount of error in the anchor positioning?

does the orientation of the anchors matter?

**Reply by krichardsson** [reply]
The best placement of the anchors is to keep them "standing", that is with the antenna up and connectors down. The measured distance between two units (anchors or Crazyflies) changes quite a lot (probalby around 40 cm) depending on the relative orientation of the antennas. The antenna radiation pattern is fairly uniform when the anchors are standing up and this is why this is better. This is generally a pretty tricky problem to solve, you can read more about it [here](https://www.bitcraze.io/author/wenda/) if you are interested.

The environment around the anchors (walls, objects, humans) also effect the measured distance and the estimated position. The problem is that it is very hard to predict what kind of effects you will get, I think trial and error is the only way I can recommend.

**Reply by Kalib-Drift** [reply]
Alright, I have accounted for most of these issues but I will try setting up the space with upright anchors.

I have already sent an email to your team but it would be great to set up a meeting to potentially go over this issue in a more efficient way.
